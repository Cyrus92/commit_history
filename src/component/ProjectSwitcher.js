import React from 'react';
import Button from 'react-bootstrap/Button';
import 'bootstrap/dist/css/bootstrap.min.css';
import 'bootstrap/dist/js/bootstrap.min.js';
import { Octokit } from "@octokit/rest"

const apiContext = "https://api.github.com";

export default class ProjectSwitcher extends React.Component {
    constructor(props) {
        super(props);
        // Set initial states
        this.state = {
            owner:"vinitshahdeo",// "cyrus92",
            repo: "Library-Management-System", //commit_history",
            totalCommits: 0,
            commits: "",
            page: 1,
            isLoading: true,
            disableShowCommitsBtn: true
        };    
        // bind Change Owner event
        this.handleChangeOwner = this.handleChangeOwner.bind(this);
        // bind Change Repo event
        this.handleChangeRepo = this.handleChangeRepo.bind(this);
        // bind Change fetchCommits event
        this.fetchCommits = this.fetchCommits.bind(this);
        // bind  fetchCommitsProcessing 
        this.fetchCommitsProcessing = this.fetchCommitsProcessing.bind(this);
        
    }

    componentDidMount() {
        // fetch commits according to default states
        this.fetchCommits(this.state.owner, this.state.repo)
        
        
    }

    // Change Owner event
    handleChangeOwner(e) {
        // set Owner field value to owner state
        this.setState({owner: e.target.value})
        // send Fetch commits request
        this.fetchCommits(e.target.value, this.state.repo)
    }

    // Change Repo event
    handleChangeRepo(e) {
        // set Repo field value to repo state
        this.setState({repo: e.target.value})
        // send Fetch commits request
        this.fetchCommits(this.state.owner, e.target.value)
    }

    
    // Process Fetch commits data
    fetchCommitsProcessing(response) {
        
        if (response.status==200) {
            const data = response.data; 
            // set totalCommits 
            this.setState({totalCommits: data.length}) 
            // set commits 
            this.setState({commits: data})
            // set disableShowCommitsBtn 
            if (this.state.totalCommits > 0) {
                this.setState({disableShowCommitsBtn: false})
            }else{
                this.setState({disableShowCommitsBtn: true})
            }
          }else{
            // set totalCommits 
            this.setState({totalCommits: 0})  
            // set commits 
            this.setState({commits: ""})
            // set disableShowCommitsBtn 
            this.setState({disableShowCommitsBtn: true})
        }
    }

    // Fetch commits
    async fetchCommits(owner="", repo="") {
        
        const octokit = new Octokit({
            auth: process.env.REACT_APP_GITHUB_TOKEN
        })
        
          
        // Request to server ends  
        await octokit.request(
            'GET /repos/{owner}/{repo}/commits', {
                owner: owner,
                repo: repo
        })
        // processing
        .then(response => this.fetchCommitsProcessing(response) )
        .catch( err => 
            // reset states 
            this.setState(
                {
                    totalCommits: 0,
                    commits: "",
                    disableShowCommitsBtn: true
                }
            )  
        );
    } 
    

    // Render component
    render() {
        return (
            <div className="row" >
                {/* Input to change owner and fetch commits */}
                <div className="col-md-5" >
                    <label className="float-left" >Owner</label>
                    <input
                        id="owner"
                        type="text"
                        placeholder="Enter owner name"
                        className="form-control"
                        onChange={this.handleChangeOwner}
                        value={this.state.owner}
                        required
                    />
                </div>
                
                <div className="col-md-7" >
                    <div className="row" >
                        {/* Input to change repo and fetch commits */}
                        <div className="col-md-8" >
                            <label className="float-left" >Repo</label><br/>
                            <input
                                id="repo"
                                type="text"
                                placeholder="Enter repo name"
                                className="form-control"
                                onChange={this.handleChangeRepo}
                                value={this.state.repo}
                                required
                            />
                        </div>
                        {/* Button to start updating commitsList */}
                        <div className="col-md-4" >
                            <br/>
                            <Button disabled={this.state.disableShowCommitsBtn} id="showCommitsBtn" variant="primary" >Show {this.state.totalCommits>0?this.state.totalCommits:""} commits</Button>
                        </div>
                    </div>
                    
                </div>
            </div>
        )
    }
}