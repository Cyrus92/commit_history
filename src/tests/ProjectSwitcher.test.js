import React from 'react';
import { shallow, mount } from 'enzyme';
import ProjectSwitcher from '../component/ProjectSwitcher';
describe("ProjectSwitcher component", () => {

    it("should render", () => {
        const wrapper = shallow(<ProjectSwitcher />);
    });

    it("should render initial layout", () => {
        // when
        const component = shallow(<ProjectSwitcher />);
        // then
        expect(component.getElements()).toMatchSnapshot();
    });

    it("should create owner, totalCommits and commits states", () => {
        // given
        const component = shallow(<ProjectSwitcher />);
        const form = component.find('#owner');
        // when
        form.props().onChange({target: {
        name: 'myName',
        value: 'myValue'
        }});
        // then
        expect(component.state('owner')).toBeDefined();
        expect(component.state('totalCommits')).toBeDefined();
        expect(component.state('commits')).toBeDefined();
    });
    
    it("should create owner state with the event value", () => {
        // given
        const component = shallow(<ProjectSwitcher />);
        const form = component.find('#owner');
        // when
        form.props().onChange({target: {
        name: 'myName',
        value: 'newValue'
        }});
        // then
        expect(component.state('owner')).toEqual('newValue');
    });

    it("should create repo, totalCommits and commits states", () => {
        // given
        const component = shallow(<ProjectSwitcher />);
        const form = component.find('#repo');
        // when
        form.props().onChange({target: {
        name: 'myName',
        value: 'myValue'
        }});
        // then
        expect(component.state('repo')).toBeDefined();
        expect(component.state('totalCommits')).toBeDefined();
        expect(component.state('commits')).toBeDefined();
    });

    
    it("should create repo state with the event value", () => {
        // given
        const component = shallow(<ProjectSwitcher />);
        const form = component.find('#repo');
        // when
        form.props().onChange({target: {
        name: 'myName',
        value: 'newValue'
        }});
        // then
        expect(component.state('repo')).toEqual('newValue');
    });

    it("should create totalCommits and commits states after calling fetchCommits function", () => {
        // given
        const component = shallow(<ProjectSwitcher />);
        const instance = component.instance();
        // when
        instance.fetchCommits();
        // then        
        expect(component.state('totalCommits')).toBeDefined();
        expect(component.state('commits')).toBeDefined();
    });

    
    

});