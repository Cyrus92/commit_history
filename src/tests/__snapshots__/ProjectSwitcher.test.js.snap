// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectSwitcher component should render initial layout 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <label
        className="float-left"
      >
        Owner
      </label>
      <input
        className="form-control"
        id="owner"
        onChange={[Function]}
        placeholder="Enter owner name"
        required={true}
        type="text"
        value="cyrus92"
      />
    </div>
    <div
      className="col-md-7"
    >
      <div
        className="row"
      >
        <div
          className="col-md-8"
        >
          <label
            className="float-left"
          >
            Repo
          </label>
          <br />
          <input
            className="form-control"
            id="repo"
            onChange={[Function]}
            placeholder="Enter repo name"
            required={true}
            type="text"
            value="commit_history"
          />
        </div>
        <div
          className="col-md-4"
        >
          <br />
          <Button
            active={false}
            disabled={true}
            id="showCommitsBtn"
            variant="primary"
          >
            Show 
            
             commits
          </Button>
        </div>
      </div>
    </div>
  </div>,
]
`;
